compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/AssertionLowering.kt

```diff
@@ -0,0 +1,177 @@
+import org.jetbrains.kotlin.ir.IrElement
+import org.jetbrains.kotlin.ir.IrStatement
+import org.jetbrains.kotlin.ir.UNDEFINED_OFFSET
+import org.jetbrains.kotlin.ir.builders.*
+import org.jetbrains.kotlin.ir.declarations.*
+import org.jetbrains.kotlin.ir.declarations.impl.IrFieldImpl
+import org.jetbrains.kotlin.ir.expressions.IrCall
+import org.jetbrains.kotlin.ir.expressions.IrExpression
+import org.jetbrains.kotlin.ir.expressions.impl.IrClassReferenceImpl
+import org.jetbrains.kotlin.ir.expressions.impl.IrGetObjectValueImpl
+import org.jetbrains.kotlin.ir.symbols.impl.IrFieldSymbolImpl
+import org.jetbrains.kotlin.ir.types.getClass
+import org.jetbrains.kotlin.ir.types.typeWith
+import org.jetbrains.kotlin.ir.util.functions
+import org.jetbrains.kotlin.ir.util.patchDeclarationParents
+import org.jetbrains.kotlin.ir.visitors.IrElementTransformer
+import org.jetbrains.kotlin.name.Name
+import org.jetbrains.kotlin.util.OperatorNameConventions
+import org.jetbrains.kotlin.utils.addToStdlib.safeAs
+
+internal val assertionPhase = makeIrFilePhase(
+    ::AssertionLowering,
+    name = "Assertion",
+    description = "Lower assert calls depending on the assertions mode"
+)
+
+private class AssertionLowering(private val context: JvmBackendContext) :
+    FileLoweringPass,
+    IrElementTransformer<AssertionLowering.ClassInfo?>
+{
+    // Keeps track of the $assertionsDisabled field, which we generate lazily for classes containing
+    // assertions when compiled with -Xassertions=jvm.
+    private inner class ClassInfo(val irClass: IrClass) {
+        private val lazyAssertionsDisabledField: Lazy<IrField> = lazy {
+            val descriptor = WrappedFieldDescriptor()
+            IrFieldImpl(
+                UNDEFINED_OFFSET, UNDEFINED_OFFSET,
+                JvmLoweredDeclarationOrigin.GENERATED_ASSERTION_ENABLED_FIELD,
+                IrFieldSymbolImpl(descriptor),
+                Name.identifier("\$assertionsDisabled"), context.irBuiltIns.booleanType, Visibilities.PUBLIC,
+                isFinal = true, isExternal = false, isStatic = true
+            ).apply {
+                descriptor.bind(this)
+                parent = irClass
+                initializer = context.createIrBuilder(irClass.symbol).run {
+                    at(this@apply)
+                    irExprBody(irNot(irCall(this@AssertionLowering.context.ir.symbols.desiredAssertionStatus).apply {
+                        dispatchReceiver = getJavaClass(irClass)
+                    }))
+                }
+            }
+        }
+
+        val assertionsDisabledField by lazyAssertionsDisabledField
+
+        val needsAssertionDisabledField: Boolean
+            get() = lazyAssertionsDisabledField.isInitialized()
+
+        fun getAssertionDisabled(irBuilder: IrBuilderWithScope): IrExpression =
+            irBuilder.irGetField(null, assertionsDisabledField)
+
+        private fun getJavaClass(irClass: IrClass): IrExpression {
```

Comment:
This looks like a third copy of this code, maybe it should be extracted into a common method of CrIrType, for example?

Reply:
I've added it as an extension method on `JvmBackendContext` in `CallableReferenceLowering`.